nixos/modules/security/systemd-sandbox.nix

```diff
@@ -0,0 +1,69 @@
+{ config, lib, ... }:
+with lib;
+{
+  options.systemd.services = with types; mkOption {
+    type = attrsOf (submodule ({ name, config', ... }: {
+      options.sandbox = mkOption {
+        description = ''
+          Level of systemd service sandboxing for this particular unit.
+          All systemd confinement options are overrridable by assigning
+          another value to allow service-specific access.
+
+          0 is no confinement at all.
+          1 is basic sandboxing, probably not breaking anything.
+          2 is the highest amount of sandboxing this module supports right now.
+        '';
+        default = 0;
+        type = enum [ 0 1 2 ];
```

Comment:
I think more descriptive strings would be better - something like
```
enum [ "unsafe" "basic" "full" ];
```
What happens when you set it to default 1 or 2 btw? :)

Reply:
Yeah, the reason for the numbers is forward compatibility. As I have seen over the past systemd releases, more confinement options are added over time (for example `ProtectKernelLogs` and `ProtectClock`). Enabling them for all services which have confinement enabled would most likely break some of them. This way, we can add more options (and numeric values) in the future without having to do someting like `[ "unsafe" "basic" "full" "fuller" "fullest" "full2023" "full_newer" ]`

I also thought about setting some of the options by default (like `PrivateTmp` since postgres has its socket in /run now), but I suspect a lot of stuff to break and I cannot test all nixos modules ;) The first that comes to mind is Docker which really needs cgroups, for example.
Also, this affects **ALL** units, including systemd-shipped units like udev and this PR would most likely break all NixOS systems on the unstable channel.

Reply:
Thanks, that makes sense. I will play with it in VM.